arangod/Metrics/ClusterMetricsFeature.cpp

```diff
@@ -67,130 +67,248 @@ void ClusterMetricsFeature::validateOptions(
   _count.store(kStop, std::memory_order_release);
-  _handle = nullptr;
+  std::atomic_store_explicit(&_timer, {}, std::memory_order_relaxed);
+  std::atomic_store_explicit(&_update, {}, std::memory_order_relaxed);
 }
 
-void ClusterMetricsFeature::asyncUpdate() {
-  if (_count.load(std::memory_order_acquire) % 2 != kStop &&
-      _count.fetch_add(kUpdate, std::memory_order_acq_rel) == 0) {
-    scheduleUpdate();
+void ClusterMetricsFeature::stop() { beginShutdown(); }
+
+std::optional<std::string> ClusterMetricsFeature::update(
+    CollectMode mode) noexcept {
+  if (mode == CollectMode::TriggerGlobal) {
+    auto const count = _count.fetch_add(kUpdate, std::memory_order_acq_rel);
+    if (count == 0) {
+      rescheduleUpdate(0);
+    }
+    return std::nullopt;
   }
+  TRI_ASSERT(mode != CollectMode::Local);
+  auto& nf = server().getFeature<NetworkFeature>();
+  auto& cf = server().getFeature<ClusterFeature>();
+  auto& ci = cf.clusterInfo();
+  try {
+    // TODO(MBkkt) Need to return Future<std::optional<std::string>>:
+    // 1) other CollectModes: invalidFuture()
+    // 2) follower:           makeFuture(leader)
+    // 3) leader:             metricsOnLeader(...).thenValue(...)
+    auto [leader, version] = ci.getMetricsState(false);
+    if (leader) {
+      return leader;
+    }
+    if (!leader && mode == CollectMode::WriteGlobal) {
+      writeData(version, metricsOnLeader(nf, cf).getTry());
+    }
+  } catch (...) {
+  }
+  return std::nullopt;
+}
+
+void ClusterMetricsFeature::rescheduleTimer() noexcept {
+  auto h = SchedulerFeature::SCHEDULER->queueDelayed(
+      RequestLane::DELAYED_FUTURE, std::chrono::seconds{_timeout},
+      [this](bool canceled) noexcept {
+        if (canceled || wasStop()) {
+          return;
+        }
+        update(CollectMode::TriggerGlobal);
+        rescheduleTimer();
+      });
+  std::atomic_store_explicit(&_timer, std::move(h), std::memory_order_relaxed);
 }
 
-void ClusterMetricsFeature::scheduleUpdate() noexcept {
-  auto now = std::chrono::steady_clock::now();
-  auto next = _lastUpdate + std::chrono::seconds{_timeout};
-  _handle = SchedulerFeature::SCHEDULER->queueDelayed(
-      RequestLane::CLUSTER_INTERNAL, (now < next ? next - now : 0ns),
-      [this](bool canceled) {
-        if (canceled) {
+void ClusterMetricsFeature::rescheduleUpdate(uint32_t timeout) noexcept {
+  auto h = SchedulerFeature::SCHEDULER->queueDelayed(
+      RequestLane::CLUSTER_INTERNAL, std::chrono::seconds{timeout},
+      [this](bool canceled) noexcept {
+        if (canceled || wasStop()) {
           return;
         }
         if (_count.exchange(kUpdate, std::memory_order_acq_rel) % 2 == kStop) {
-          // If someone call more than 1 billion asyncUpdate() before we execute
-          // store we defer Stop to next try. But it's impossible,
-          // so it's valid optimization: exchange + store instead of cas loop
+          // If someone call more than billion update(TriggerGlobal)
```

Comment:
> I do not understand this comment.

This comment about very rare situation then we want decrease size of virtual queue (_count) and in the same time in parallel was call beginShutdown. In that case we can go under this if condition. And we should restore _flags stop value.
This comment means if someone call more than billion update(TriggerGlobal) between exchange and store, here will be bug, but we assume it's never possible. so we can use exchange + store instead of fetch_or (cas)

> This seems to be a case of severe overengineering.

It simple pattern, maybe put flags and virtual queue to single atomic a little difficult, but I think it's better